libsolidity/ast/Types.cpp

```diff
@@ -3445,13 +3445,15 @@ FunctionTypePointer FunctionType::asExternallyCallableFunction(bool _inLibrary,
 
 	TypePointers parameterTypes;
 	for (auto const& t: m_parameterTypes)
-	{
-		auto refType = dynamic_cast<ReferenceType const*>(t);
-		if (refType && refType->location() == DataLocation::CallData)
-			parameterTypes.push_back(TypeProvider::withLocation(refType, DataLocation::Memory, true));
-		else
-			parameterTypes.push_back(t);
-	}
+		parameterTypes.push_back(
+			TypeProvider::withLocationIfReference(DataLocation::CallData, DataLocation::Memory, t, true)
```

Comment:
Shouldn't `t` be the first argument / parameter?

Reply:
Also, it is really difficult to understand unless you really know what the function does. What about being a bit more explicit and doing it in two-steps with the first step being `ReferenceType::isReferenceWithLocation(t, Calldata)` or something like that?